src/bounded_queue.ml

```diff
@@ -0,0 +1,113 @@
+(*
+  structure node t value: data type, next: pointer to node t
+  structure queue t Head: pointer to node t, Tail: pointer to node t, H lock: lock type, T lock: lock type
+*)
+
+type 'a node = Nil | Next of 'a * 'a node ref
+
+type 'a t = {
+  mutable head : 'a node ref;
+  mutable tail : 'a node ref;
+  head_lock : Mutex.t;
+  tail_lock : Mutex.t;
+  not_empty : Condition.t;
+  not_full : Condition.t;
+  size : int Atomic.t;
+  capacity : int;
+}
+
+(* create and initialize the 2 lock queue *)
+let create () =
+  let dummy = ref Nil in
+  {
+    head = dummy;
+    tail = dummy;
+    head_lock = Mutex.create ();
+    tail_lock = Mutex.create ();
+    not_empty = Condition.create ();
+    not_full = Condition.create ();
+    size = Atomic.make 0;
+    capacity = 1_000_000;
+  }
+
+(* push element to tail of queue *)
+let push t value =
+  let new_tail = ref Nil in
+  let new_node = Next (value, new_tail) in
+  Mutex.lock t.tail_lock;
+  (* Ensures queue is not full *)
+  while Atomic.get t.size = t.capacity do
+    Condition.wait t.not_full t.tail_lock
+  done;
+  (* Check if queue is empty before this insertion *)
+  t.tail := new_node;
+  t.tail <- new_tail;
+  let empty = Atomic.fetch_and_add t.size 1 = 0 in
+  Mutex.unlock t.tail_lock;
+  (* If it was empty then signal the waiting pop threads to wake up *)
+  if empty then (
+    Mutex.lock t.head_lock;
+    Condition.broadcast t.not_empty;
+    Mutex.unlock t.head_lock)
+
+(* pop element from head of queue *)
+let pop t =
+  Mutex.lock t.head_lock;
+  (* Ensures queue is not empty *)
+  while Atomic.get t.size = 0 do
+    Condition.wait t.not_empty t.head_lock
+  done;
+  (* Check if queue is full before this deletion *)
+  let popped =
+    match !(t.head) with
+    | Nil -> assert false
+    | Next (value, next) ->
+        t.head <- next;
+        value
+  in
+  let full = Atomic.fetch_and_add t.size (-1) = t.capacity in
+  Mutex.unlock t.head_lock;
+  (* It it was full then signal the waiting push threads to wake up *)
+  if full then (
+    Mutex.lock t.tail_lock;
+    Condition.broadcast t.not_full;
```

Comment:
Is it necessary to wake up all the push threads?

Reply:
if you don't wake up all respective operation (push/pop) threads then it could result in lost wakeup problem. it's described in section 8.2.2 in art of multiprocessor programming. Essentially one of the threads could miss the wakeup call and need to wait forever. Even just comparing performance in benchmarks using broadcast seems to do better than signal by a slight margin. (Note: the program won't terminate in the scenarios when the lost wakeup problem occurs).

Reply:
Hmm... I don't have that book and the chapter seems to discuss the topic in general.  Yes, it is possible to use mutexes and conditions in such a way that broadcast is necessary (even if only wait can actually proceed).  However, I don't immediately see how that would happen in this particular case.  Each of the two conditions is `wait`ed for only one piece of state and the state being checked is exactly the one modified prior to corresponding `broadcast`.  I must be missing something.  Could you describe a sequence of operations that causes a lost wakeup?

Reply:
Thinking about this now there is a kind of future issue with this implementation: cancellation.  This doesn't apply yet, because the code uses the Stdlib `Mutex` and `Condition`, which do not support cancellation and are also otherwise scheduler unfriendly (we really shouldn't use those).

Concurrent data structures should ideally be scheduler friendly such that they do not block the thread / domain from running other fibers.  Data structures should also be safe with respect to cancellation such that they do not leave the data structure (or other fibers operating on the data structure) in an invalid state after cancellation.

Assume, for a moment, that Mutex and Condition would actually support cancellation (such as the Mutex and Condition in Eio), and consider the following (code from `pop` with additional comments):

```ocaml
  (* The fiber has been cancelled at this point. *)
  
  let full = Atomic.fetch_and_add t.size (-1) in
  Mutex.unlock t.head_lock;
  
  (* It it was full then signal the waiting push threads to wake up *)
  
  if full = t.capacity then (
    Mutex.lock t.tail_lock;  (* <- As fiber was cancelled, this will raise the cancellation exception... *)
    Mutex.unlock t.tail_lock;
    Condition.broadcast t.not_full);  (* ...and no wakeup is broadcast *)
```

A way to fix this is to temporarily prevent cancellation, but unfortunately we don't yet have scheduler agnostic and friendly mutexes and conditions nor a [scheduler agnostic way to prevent cancellation](https://github.com/polytypic/io#cancellation).

So, I'd recommend to add TODO notes to the code:
- TODO: Use scheduler friendly mutexes and condition variables
- TODO: Make operations cancellation safe

Pinging also @lyrm as this is a general future issue with lock based data structures.